README
NAME
    Spunge::DB - Makes fork() in debugger to open a new Tmux window

VERSION
    This documentation refers to the module contained in the distribution
    Spunge-DB version 0.1.0.

SYNOPSIS
    As a helper for the debugger, the module should be used this way:

        perl -MSpunge::DB -d your_script.pl

    You should run it from inside the "tmux" window manager.

DESCRIPTION
    The Perl's standard debugger requires additional stuff when the debugged
    Perl program use the fork() built-in.

    This module is about to solve the trouble which visible like this:

      ######### Forked, but do not know how to create a new TTY. #########
      Since two debuggers fight for the same TTY, input is severely entangled.

      I know how to switch the output to a different window in xterms, OS/2
      consoles, and Mac OS X Terminal.app only.  For a manual switch, put the
      name of the created TTY in $DB::fork_TTY, or define a function
      DB::get_fork_TTY() returning this.

      On UNIX-like systems one can get the name of a TTY for the given window
      by typing tty, and disconnect the shell from TTY by sleep 1000000.

    "OS/2" is a fun for me but in the past. "Mac OS X" is a more chance but
    still isn't real. And for "xterm" ... who wants to keep it on your
    server? Gimme some stones to throw on that one.

    But the pseudo-terminal device isn't much about "GUI"s by its nature so
    the problem behind the bars of the perl5db.pl report ( see more detailed
    problem description at the PerlMonks thread
    <http://perlmonks.org/?node_id=128283> ) is the consoles management.
    It's a kind of a tricky, for example, to start the next "ssh" session
    initiated from the machine serving as an "sshd" server for the existing
    session.

    Thus we kind of have to give a chance to the consoles management with a
    software capable to run on a server machine without as much dependencies
    as an "xterm". This module is a try to pick the Tmux
    <http://tmux.sf.net> windows manager for such a task.

SUBROUTINES/METHODS
    All of the following are functions:

  PUBLIC
   "DB::get_fork_TTY()"
    Finds new "TTY" for the "fork()"ed process.

    Takes no arguments. Returns "Str" name of the "tty" device of the
    <tmux>'s new window created for the debugger's new process.

    Sets the $DB::fork_TTY to the same "Str" value.

  PRIVATE
   "_spawn_tty()"
    Creates a "TTY" device and returns "Str" its name.

   "_tmux_new_window()"
    Creates a given "tmux" window and returns "Str" its id/number.

   "_tmux_window_tty( $window_id )"
    Finds a given "tmux" window's tty name and returns its "Str" name based
    on a given window id/number typically from "_tmux_new_window()".

   "_read_from_cmd( $cmd => @args )"
    Takes the list containing the "Str" system() command and "Array" its
    arguments and executes it. Reads Str the output and returns it. Throws
    if no output or if the command failed.

   "_croak_on_cmd( $cmd => @args, $happen )"
    Takes the "Str" command, "Array" its arguments and "Str" the reason of
    its failure, examines the $? and dies with explanation on the system()
    command failure.

CONFIGURATION AND ENVIRONMENT
    The module requires the Tmux <http://tmux.sf.net> window manager for the
    console to be present in the system.

    Configuration is made via environment variables, the default is taken
    for each of them with no such variable is set in the environment:

  "SPUNGE_TMUX_FQDN"
    The "tmux" binary name with the full path.

    Default : "/usr/local/bin/tmux"

  "SPUNGE_TMUX_CMD_NEWW"
    The system() arguments for a "tmux" command for opening a new window and
    with output of a window address from "tmux". String is splitted by
    spaces to be a list of parameters.

    Dafeult : "neww -P"

  "SPUNGE_TMUX_CMD_NEWW_EXEC"
    The system() or a shell command to be given to the
    "SPUNGE_TMUX_CMD_NEWW" command to be executed in a brand new created
    wiondow. It should wait unexpectedly and do nothing till the debugger
    catches the device and puts in into the proper use.

    Dafeult : "sleep 1000000"

  "SPUNGE_TMUX_CMD_TTY"
    Command- line parameter(s) for a "tmux" command to find a "tty" name in
    the output. The string is splitted then by spaces. The "tmux"'s window
    address is added then as the very last argument.

    Dafeult : "lsp -F #{pane_tty} -t"

DIAGNOSTICS
    The command ...
        Typically the error message starts with the command the Spunge::DB
        tried to execute, including the command's arguments.

    failed opening command: ...
        The command was not taken by the system as an executable binary
        file.

    ... didn't write a line
    failed reading command: ...
        Command did not output exactly one line of the text.

    ... did not finish
        Command outputs more than one line of the text.

    provided empty string
        Command outputs exactly one line of the text and the line is empty.

    failed to execute: ...
        There was failure executing the command

    child died with(out) signal X, Y coredump
        Command was killed by the signal X and the coredump is (not) located
        in Y.

    child exited with value X
        Command was not failed but there are reasons to throw an error like
        the wrong command's output.

DEPENDENCIES
    * "Perl 5.6.0+" is available from The Perl website <http://www.perl.org>

    * Config, Cwd, DB, ExtUtils::MakeMaker, File::Find, File::Spec are
    available in core "Perl" distribution version 5.6.0 and later

    * Const::Fast is available from "CPAN"

    * Module::Build is available in core "Perl" distribution since version
    5.9.4

    * Scalar::Util is available in core "Perl" distribution since version
    5.7.3

    * Sort::Versions is available from "CPAN"

    * Test::Exception is available from "CPAN"

    * Test::More is available in core "Perl" distribution since version
    5.6.2

    * Test::Most is available from "CPAN"

    * Test::Strict is available from "CPAN"

    * autodie is available in core "Perl" distribution since version 5.10.1

    * "Tmux" v1.6+ is available from The Tmux website
    <http://tmux.sourceforge.net>

    Most of them can easily be found in your operating system
    distribution/repository.

BUGS AND LIMITATIONS
    You can make new bug reports, and view existing ones, through the web
    interface at <http://bugs.vereshagin.org/product/Spunge-DB>.

WEB SITE
    The web site of Spunge::DB
    <http://gitweb.vereshagin.org/Spunge-DB/README.html> currently consists
    of only one page cause it's a very small module.

SUPPORT
  Perldoc
    You can find documentation for this module with the perldoc command.

      perldoc Spunge::DB

  Email
    You can email the author of this module at "peter@vereshagin.org" asking
    for help with any problems you have.

  Bugs / Feature Requests
    Please report any bugs or feature requests by email to
    "peter@vereshagin.org", or through the web interface at
    <http://bugs.vereshagin.org/product/Spunge-DB>. You will be
    automatically notified of any progress on the request by the system.

  Source Code
    The code is open to the world, and available for you to hack on. Please
    feel free to browse it and play with it, or whatever. If you want to
    contribute patches, please send me a diff or prod me to pull from your
    repository :)

    <http://gitweb.vereshagin.org/Spunge-DB>

      git clone https://github.com/petr999/Spunge-DB.git

AUTHOR
    Peter Vereshagin <http://vereshagin.org> <peter@vereshagin.org>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2012 by Peter Vereshagin.

    This is free software, licensed under:

      The (three-clause) BSD License

DISCLAIMER OF WARRANTY
    BECAUSE THIS SOFTWARE IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
    FOR THE SOFTWARE, TO THE EXTENT PERMITTED BY APPLICABLE LAW. EXCEPT WHEN
    OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
    PROVIDE THE SOFTWARE "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER
    EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
    WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE. THE
    ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE SOFTWARE IS WITH
    YOU. SHOULD THE SOFTWARE PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL
    NECESSARY SERVICING, REPAIR, OR CORRECTION.

    IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
    WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
    REDISTRIBUTE THE SOFTWARE AS PERMITTED BY THE ABOVE LICENCE, BE LIABLE
    TO YOU FOR DAMAGES, INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL, OR
    CONSEQUENTIAL DAMAGES ARISING OUT OF THE USE OR INABILITY TO USE THE
    SOFTWARE (INCLUDING BUT NOT LIMITED TO LOSS OF DATA OR DATA BEING
    RENDERED INACCURATE OR LOSSES SUSTAINED BY YOU OR THIRD PARTIES OR A
    FAILURE OF THE SOFTWARE TO OPERATE WITH ANY OTHER SOFTWARE), EVEN IF
    SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
    DAMAGES.



Changes
Revision History for Spunge-DB
==============================

{{$NEXT}}

0.0.9     2012-10-12 17:48:07 Europe/Moscow

v0.0.7 2012-10-12
-----------------

2012-10-12 14:42:59 +0400
PEADME.pod version bump

2012-10-12 14:27:57 +0400
More files on the block


v0.0.6 2012-10-12
-----------------

2012-10-12 13:17:39 +0400
Removed timestamp from README.html

2012-10-12 13:14:25 +0400
Removed timestamp from README.html


v0.0.5 2012-10-12
-----------------

2012-10-12 11:28:38 +0400
README.html removed due to having a timestamp

2012-10-12 11:16:26 +0400
added READMEs

2012-10-12 11:02:09 +0400
Files of a build in the master branch

2012-10-12 10:54:10 +0400
README.pod version bump

2012-10-12 10:39:36 +0400
added README.pod


v0.0.4 2012-10-12
-----------------

2012-10-12 02:13:27 +0400
Ticket 30 (82676ebc-d911-43d7-96d9-8c9cbc906508)
   set up web links:

      - GitHub repository of the Spunge-DB project;

      - GitWeb page  of the Spunge-DB project at a Vereshagin.Org;

      - Bugzilla page of the Spunge-DB project at a Vereshagin.Org, and in
    a
       BUGS AND LIMITATIONS section, too.


v0.0.3 2012-10-07
-----------------

2012-10-07 06:49:34 +0400
more POD fixes

2012-10-07 06:27:10 +0400
POD fixes
   Ticket 11 (fb4da310-a328-4fa3-b8b7-0185e95919a7) Smaller POD fixes:

      - build POD SYNOPSIS section. Disable DZP:PodVersion in favor of
       PodWeaver;

      - ' $DB::fork_TTY to the same <Str> ' should be enclosed in C<>;

      - AUTHOR section should include http://vereshagin.org;

      - Function names in the POD headers should be supplied the braces();

      - Build CONFIGURATION AND ENVIRONMENT section in the POD;

      - Some of the  =sub headers lack '>' in favor of E<gt>.

    Ticket 14 (0ec84f72-560e-4850-a31f-34b90518b338) DEPENDENCIES section:

      - Make a DEPENDENCIES section in the POD in a PBP way.

    Ticket 17 (127641d7-69f1-4c5c-b937-3866c7ecd76b) write croak messages
    to DIAGNOSTICS POD

    DIAGNOSTICS POD should be filled with croak_on_cmd() messages.

    Ticket 19 (85782e29-4094-4cb4-9cf4-c9be38b5fa8a) Point POD links from
    perl built-ins to the corresponding perlfunc sections. Make them work
    in the resulting HTML:

      - L<system>;

      - L<fork>.

    all of them are perl built-in subroutines.


v0.0.2 2012-10-07
-----------------

2012-10-07 01:44:03 +0400
WORKSFORME
   Working code is created:

      - the module Spunge::DB with subs, docs, etc.
     - tests include:

        - 10-use.t tries to use the module. No bad if it only require()s
    the
         module instead of use() as module does not @EXPORT anything;
       - 20-cmd.t tries the basic commands and the module's built-in
    exception
         catching;
       - 30-tmux.t tries to call functions executing the Tmux.

2012-10-04 21:58:05 +0400
Added 'use' and 'strict' test

2012-10-04 18:40:31 +0400
Added synopsis to POD.


===================================
End of changes in the last 365 days
===================================


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Config" => 0,
    "File::Find" => 0,
    "Module::Build" => "0.3601",
    "Scalar::Util" => 0,
    "Sort::Versions" => 0,
    "Test::Exception" => 0,
    "Test::More" => "0.94",
    "Test::Most" => 0,
    "Test::Strict" => 0,
    "autodie" => 0,
    "perl" => "5.004"
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Makes fork() in debugger to open a new Tmux window",
  "dist_author" => [
    "L<Peter Vereshagin|http://vereshagin.org> <peter\@vereshagin.org>"
  ],
  "dist_name" => "Spunge-DB",
  "dist_version" => "0.1.0",
  "license" => "bsd",
  "module_name" => "Spunge::DB",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Const::Fast" => 0,
    "Cwd" => 0,
    "File::Spec" => 0,
    "perl" => "5.006",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Spunge/DB.pm
# ABSTRACT: Makes fork() in debugger to open a new Tmux window
package Spunge::DB;

# VERSION

# Helps you to behave
use strict;
use warnings;

### MODULES ###
#
# Glues up path components
use File::Spec;

# Resolves up symlinks
use Cwd;

# Dioes in a nicer way
use Carp;

# Reads configuration
use Spunge::DB::Config;

# Makes constants possible
use Const::Fast;

### CONSTANTS ###
#

### SUBS ###
#
# Function
# Gets the tty name, sets the $DB::fork_TTY to it and returns it.
# Takes     :   n/a
# Requires  :   DB, Spunge::DB
# Overrides :   DB::get_fork_TTY()
# Changes   :   $DB::fork_TTY
# Returns   :   Str tty name $DB::fork_TTY
sub DB::get_fork_TTY {

    # Create a TTY
    my $tty_name = Spunge::DB::_spawn_tty();

    # Output the name both to a variable and to the caller
    no warnings qw/once/;
    $DB::fork_TTY = $tty_name;
    return $tty_name;
}

# Function
# Spawns a TTY and returns its name
# Takes     :   n/a
# Returns   :   Str tty name
sub _spawn_tty {

    # Create window and get its tty name
    my $window_id = _tmux_new_window();
    my $tty_name  = _tmux_window_tty($window_id);

    return $tty_name;
}

# Function
# Creates new 'tmux' window  and returns its id/number
# Takes     :   n/a
# Depends   :   On 'tmux_fqdn', 'tmux_neww', 'tmux_neww_exec' configuration
#               parameters
# Returns   :   Str id/number of the created 'tmux' window
sub _tmux_new_window {
    my @cmd_to_read = (
        Spunge::DB::Config->get_config('tmux_fqdn'),
        split( /\s+/, Spunge::DB::Config->get_config('tmux_cmd_neww') ),
        Spunge::DB::Config->get_config('tmux_cmd_neww_exec'),
    );

    my $window_id = _read_from_cmd(@cmd_to_read);

    return $window_id;
}

# Function
# Gets a 'tty' name from 'tmux's window id/number
# Takes     :   Str 'tmux' window id/number
# Depends   :   On 'tmux_fqdn', 'tmux_cmd_tty' configuration parameters
# Returns   :   Str 'tty' device name of the 'tmux' window
sub _tmux_window_tty {
    my $window_id = shift;

    # Concatenate the 'tmux' command and read its output
    my @cmd_to_read = (
        Spunge::DB::Config->get_config('tmux_fqdn'),
        split( /\s+/, Spunge::DB::Config->get_config('tmux_cmd_tty') ),
        $window_id,
    );
    my @tmux_cmd = (@cmd_to_read);
    my $tty_name = _read_from_cmd(@tmux_cmd);

    return $tty_name;
}

# Function
# Reads the output of a command supplied with parameters as the argument(s)
# and returns its output.
# Takes     :   Array[Str] command and its parameters
# Throws    :   If command failed or the output is not the non-empty Str
#               single line
# Returns   :   Output of the command supplied with parameters as arguments
sub _read_from_cmd {
    my @cmd_and_args = @_;

    # Open the pipe to read
    _croak_on_cmd( @cmd_and_args, "failed opening command: $!" )
        unless open my $cmd_output_fh => '-|',
        @cmd_and_args;

    # Read a line from the command
    _croak_on_cmd( @cmd_and_args, "didn't write a line" )
        unless defined($cmd_output_fh)
        and ( 0 != $cmd_output_fh )
        and my $cmd_out = <$cmd_output_fh>;

    # If still a byte is readable then die as the file handle should be
    # closed already
    my $read_rv = read $cmd_output_fh => my $buf, 1;
    _croak_on_cmd( @cmd_and_args, "failed reading command: $!" )
        unless defined $read_rv;
    _croak_on_cmd( @cmd_and_args, "did not finish" ) unless 0 == $read_rv;

    # Die on empty output
    chomp $cmd_out;
    _croak_on_cmd( @cmd_and_args, "provided empty string" )
        unless length $cmd_out;

    return $cmd_out;
}

# Function
# Croaks nicely on the command with an explanation based on arguments and $?
# Takes     :   Array[Str] system command, its arguments, and an explanation
#               on the situation when the command failed
# Requires  :   Carp
# Depends   :   On $? global variable set by system command failure
# Throws    :   Always
# Returns   :   n/a
sub _croak_on_cmd {
    my @cmd_args_msg = @_;

    if ( defined $? ) {
        my $msg = '';

        # Depending on $?, add it to the death note
        # Command may be a not-executable
        if ( $? == -1 ) {
            $msg = "failed to execute: $!";
        }

        # Command can be killed
        elsif ( $? & 127 ) {
            $msg = sprintf "child died with signal %d, %s coredump",
                ( $? & 127 ), ( $? & 128 ) ? 'with' : 'without';
        }

        # Command may return the exit code for clearance
        else {
            $msg = sprintf "child exited with value %d", $? >> 8;
        }

        # And the message can be returned as an appendix to the original
        # arguments
        push @cmd_args_msg, $msg;
    }

    # Report the datails via the Carp
    my $croak_msg = "The command " . join ' ' => @cmd_args_msg;
    croak($croak_msg);
}

# Returns true to require()
1;

__END__

=pod

=head1 SYNOPSIS

As a helper for the debugger, the module should be used this way:

    perl -MSpunge::DB -d your_script.pl

You should run it from inside the C<tmux> window manager.

=head1 DESCRIPTION

The Perl's standard debugger requires additional stuff when the debugged
Perl program use the L<fork()|perlfunc/fork> built-in.

This module is about to solve the trouble which visible like this:

  ######### Forked, but do not know how to create a new TTY. #########
  Since two debuggers fight for the same TTY, input is severely entangled.

  I know how to switch the output to a different window in xterms, OS/2
  consoles, and Mac OS X Terminal.app only.  For a manual switch, put the
  name of the created TTY in $DB::fork_TTY, or define a function
  DB::get_fork_TTY() returning this.

  On UNIX-like systems one can get the name of a TTY for the given window
  by typing tty, and disconnect the shell from TTY by sleep 1000000.

C<OS/2> is a fun for me but in the past. C<Mac OS X> is a more chance but
still isn't real. And for C<xterm> ... who wants to keep it on your server?
Gimme some stones to throw on that one.

But the pseudo-terminal device isn't much about C<GUI>s by its nature so the
problem behind the bars of the L<perl5db.pl> report ( see more detailed
problem description at the L<PerlMonks
thread|http://perlmonks.org/?node_id=128283> ) is the consoles management.
It's a kind of a tricky, for example, to start the next C<ssh> session
initiated from the machine serving as an C<sshd> server for the existing
session.

Thus we kind of have to give a chance to the consoles management with
a software capable to run on a server machine without as much dependencies
as an C<xterm>. This module is a try to pick the L<Tmux|http://tmux.sf.net>
windows manager for such a task.

=head1 SUBROUTINES/METHODS

All of the following are functions:

=pubsub C<DB::get_fork_TTY()>

Finds new C<TTY> for the C<fork()>ed process.

Takes no arguments. Returns C<Str> name of the C<tty> device of the <tmux>'s
new window created for the debugger's new process.

Sets the C<$DB::fork_TTY> to the same C<Str> value.

=sub C<_spawn_tty()>

Creates a C<TTY> device and returns C<Str> its name.

=sub C<_tmux_new_window()>

Creates a given C<tmux> window and returns C<Str> its id/number.

=sub C<_tmux_window_tty( $window_id )>

Finds a given C<tmux> window's tty name and returns its C<Str> name based on
a given window id/number typically from L</_tmux_new_window()>.

=sub C<_read_from_cmd( $cmd =E<gt> @args )>

Takes the list containing the C<Str> L<system()|perlfunc/system> command and
C<Array> its arguments and executes it. Reads Str the output and returns it.
Throws if no output or if the command failed.

=sub C<_croak_on_cmd( $cmd =E<gt> @args, $happen )>

Takes the C<Str> command, C<Array> its arguments and C<Str> the reason of
its failure, examines the C<$?> and dies with explanation on the
L<system()|perlfunc/system> command failure.

=head1 DIAGNOSTICS

=over

=item The command ...

Typically the error message starts with the command the L<Spunge::DB> tried
to execute, including the command's arguments.

=item failed opening command: ...

The command was not taken by the system as an executable binary file.

=item ... didn't write a line

=item failed reading command: ...

Command did not output exactly one line of the text.

=item ... did not finish

Command outputs more than one line of the text.

=item provided empty string

Command outputs exactly one line of the text and the line is empty.

=item failed to execute: ...

There was failure executing the command

=item child died with(out) signal X, Y coredump

Command was killed by the signal X and the coredump is (not) located in Y.

=item child exited with value X

Command was not failed but there are reasons to throw an error like the
wrong command's output.

=back


=head1 DEPENDENCIES

* C<Perl 5.6.0+>
is available from L<The Perl website|http://www.perl.org>

* L<Config>, L<Cwd>, L<DB>, L<ExtUtils::MakeMaker>, L<File::Find>,
L<File::Spec> are available in core C<Perl> distribution version 5.6.0 and
later

* L<Const::Fast>
is available from C<CPAN>

* L<Module::Build>
is available in core C<Perl> distribution since version 5.9.4

* L<Scalar::Util>
is available in core C<Perl> distribution since version 5.7.3

* L<Sort::Versions>
is available from C<CPAN>

* L<Test::Exception>
is available from C<CPAN>

* L<Test::More>
is available in core C<Perl> distribution since version 5.6.2

* L<Test::Most>
is available from C<CPAN>

* L<Test::Strict>
is available from C<CPAN>

* L<autodie>
is available in core C<Perl> distribution since version 5.10.1

* C<Tmux> v1.6+
is available from L<The Tmux website|http://tmux.sourceforge.net>

Most of them can easily be found in your operating system
distribution/repository.

=head1 CONFIGURATION AND ENVIRONMENT

The module requires the L<Tmux|http://tmux.sf.net> window manager for the
console to be present in the system.

Configuration is made via environment variables, the default is taken for
each of them with no such variable is set in the environment:

=head2 C<SPUNGE_TMUX_FQDN>

The C<tmux> binary name with the full path.

Default :   C</usr/local/bin/tmux>

=head2 C<SPUNGE_TMUX_CMD_NEWW>

The L<system()|perlfunc/system> arguments for a C<tmux>
command for opening a new window and with output of a window address from
C<tmux>. String is splitted by spaces to be a list of parameters.

Dafeult :  C<neww -P>

=head2 C<SPUNGE_TMUX_CMD_NEWW_EXEC>

The L<system()|perlfunc/system> or a shell command to be given to the
C<SPUNGE_TMUX_CMD_NEWW> command to be executed in a brand new created
wiondow. It should wait unexpectedly and do nothing till the debugger
catches the device and puts in into the proper use.

Dafeult :  C<sleep 1000000>

=head2 C<SPUNGE_TMUX_CMD_TTY>

Command- line  parameter(s) for a  C<tmux> command to find a C<tty> name in
the output. The string is splitted then by spaces. The C<tmux>'s window
address is added then as the very last argument.

Dafeult :  C<lsp -F #{pane_tty} -t>

=head1 WEB SITE

The web site of
L<Spunge::DB|http://gitweb.vereshagin.org/Spunge-DB/README.html> currently
consists of only one page cause it's a very small module.

=cut


lib/Spunge/DB/Config.pm
# ABSTRACT: Configuration system for Spunge::DB
package Spunge::DB::Config;

# VERSION

# Helps you to behave
use strict;
use warnings;

### MODULES ###
#
# Glues up path components
use File::Spec;

# Resolves up symlinks
use Cwd;

# Dioes in a nicer way
use Carp;

# Makes constants possible
use Const::Fast;

### CONSTANTS ###
#

# Default path to the 'tmux' binary
# Requires  :   Cwd
const my $_DEFAULT_TMUX_PATH => Cwd::realpath('/usr/local/bin');

# Default 'tmux' binary fqdn
# Depends   :   On $_DEFAULT_TMUX_PATH
# Requires  :   File::Spec
const my $_DEFAULT_TMUX_FQDN =>
    File::Spec->catfile( $_DEFAULT_TMUX_PATH => 'tmux' );

# Keep the configuration variables
my %_CONF;

# Tmux file name with full path
$_CONF{'tmux_fqdn'} = $_DEFAULT_TMUX_FQDN;

# Tmux 'neww' parameter for a system/shell command
$_CONF{'tmux_cmd_neww_exec'} = 'sleep 1000000';

# Tmux  'neww' command paraneters to be sprintf()'d with 'tmux_fqdn' and
# pushed after split by spaces the 'tmux_cmd_neww_exec' into list of
# parameters
$_CONF{'tmux_cmd_neww'} = "neww -P";

# Tmux command parameters to get a tty name
$_CONF{'tmux_cmd_tty'} = 'lsp -F #{pane_tty} -t';

# Take config override from %ENV
# Depends   :   On %ENV global of the main::
foreach my $key ( keys %_CONF ) {

    # Key for %ENV
    my $env_key = "SPUNGE_" . uc $key;

    next unless defined $ENV{$env_key};

    $_CONF{$key} = $ENV{$env_key};
}

const %_CONF => %_CONF;

### ATTRIBUTES ###
#

### SUBS ###
#

# Static method
# Returns Str argument configured as a key supplied as an argument
# Takes     :   Str argument to read config for
# Depends   :   On %_CONF package lexical
# Requires  :   Carp
# Throws    :   If no configuration found for an argument
# Returns   :   %_CONF element for an argument
sub get_config {
    shift;
    my $key = shift;

    croak("Undefined in a configuration: $key") unless defined $_CONF{$key};

    return $_CONF{$key};
}

# Static method
# Takes     :   n/a
# Depends   :   On %_CONF package lexical
# Returns   :   Array keys of %_CONF package lexical
sub get_all_config_keys { return keys %_CONF }

# Returns true to require()
1;

__END__

=pod

=head1 OVERVIEW

Most of the things are explained in L<Spunge::DB|Spunge::DB/CONFIGURATION
AND ENVIRONMENT>.

=head1 SYNOPSIS

To read the configuration L<Spunge::DB> can do this:

    use Spunge::DB;

    my $tmux_fqdn = Spunge::DB->config( 'tmux_fqdn' );

=head1 DESCRIPTION

This module reads description from environment variables and use defaults if
those are not set.

For example C<tmux_fqdn> can be overridden with C<SPUNGE_TMUX_FQDN>
variable, and so on..

=head1 SUBROUTINES/METHODS

All of the following are static methods:

=pubsub C<get_config( Str the name of the option )>

Retrieves configuration stored in an internal C<Spunge::DB::Config>
constants.

Returns Str value of the configuration parameter.

=sub C<get_all_config_keys()>

Returns Array[Str] names of all the configuration parameters.

=cut

=head1 DIAGNOSTICS

=over

=item C<Undefined in a configuration: E<lt>keyE<gt>>

Dies if no key asked was found in the configuration.

=back

=head1 CONFIGURATION AND ENVIRONMENT

See L<Spunge::DB/CONFIGURATION AND ENVIRONMENT>.

=cut



Makefile.PL

use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Makes fork() in debugger to open a new Tmux window",
  "AUTHOR" => "L<Peter Vereshagin|http://vereshagin.org> <peter\@vereshagin.org>",
  "BUILD_REQUIRES" => {
    "Config" => 0,
    "File::Find" => 0,
    "Module::Build" => "0.3601",
    "Scalar::Util" => 0,
    "Sort::Versions" => 0,
    "Test::Exception" => 0,
    "Test::More" => "0.94",
    "Test::Most" => 0,
    "Test::Strict" => 0,
    "autodie" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "Spunge-DB",
  "EXE_FILES" => [],
  "LICENSE" => "bsd",
  "NAME" => "Spunge::DB",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Const::Fast" => 0,
    "Cwd" => 0,
    "File::Spec" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.1.0",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



